docs/proposals/002-refresh-certificates.md

```diff
@@ -0,0 +1,374 @@
+node. Refreshing the certificates of all cluster nodes poses risks and possibly
+unpredictable control plane or workload downtime, therefore is out of scope for
+the current proposal.
+
+### Microcluster certificates
+
+microcluster does not provide any mechanism to rotate or refresh the cluster and
+server certificates of the nodes. Certificates for each node are created with a
+10 year TTL, but there might be environments where this is not allowed.
+
+### Refresh CA certificate
+
+Refreshing the individual node certificates can often be performed "online", as
+it typically only requires a restart of the Kubernetes services on the node
+(such that the new certificates are picked up). It often comes without much
+noticeable user-facing control plane disruption, and does not have any effect on
+the cluster workloads.
+
+On the contrary, refreshing the CA certificates of a cluster comes with serious
+downtime:
+
+- Client authentication is broken before all nodes are aware of the new CA.
+- The Kubernetes CA certificate is visible and used by workloads that must
+  interact with the Kubernetes API (e.g. CNI, controllers, etc). These pods
+  must then be restarted after the certificates have been refreshed.
+
+Due to the very disruptive nature of the operation, refreshing the cluster CA
+is kept intentionally out of scope for the existing proposal, and will be dealt
+with separately in the future.
+
+### External CA certificate
+
+When deploying Canonical Kubernetes with an external CA certificate, the cluster
+does not know the CA key. Therefore, it is unable to sign new node certificates.
+
+Due to this, it is the responsibility of the orchestration tool that is used to
+sign the new certificates and make them available on the cluster nodes.
+
+Automating any part of this manual process is out of scope for this proposal.
+
+# Implementation Details
+
+## API Changes
+<!--
+This section MUST mention any changes to the k8sd API, or any additional API
+endpoints (and messages) that are required for this proposal.
+
+Unless there is a particularly strong reason, it is preferable to add new v2/v3
+APIs endpoints instead of breaking the existing APIs, such that API clients are
+not affected.
+-->
+
+Certificate refreshes behave differently for control plane and worker nodes.
+Control plane nodes have access to the CA certificate and private key, therefore
+can easily refresh their own certificates. Worker nodes do not have access to
+the CA private key, so certificate refreshes need to happen in two steps.
+
+To accomodate this need, while also ensuring that the `k8s refresh-certs` can be
+straightforward to implement, two API calls are introduced, which clients should
+call in order.
+
+### `POST /refresh-certs/plan`
+
+```go
+type RefreshCertificatesPlanRequest struct {
+}
+
+type RefreshCertificatesPlanResponse struct {
+  // Seed needs to be passed to `/refresh-certs/run` to ensure that CSR
+  // names are valid.
+  Seed int
```

Comment:
I see that `seed` is mentioned in the CLI changes section:
```
k8s kubectl certificate approve k8sd-seed-worker-kubelet-client
```
So, I guess this is an identifier that the node knows, right? Does it need to be something specific, like a number we certainly know how to generate using available data, or a randomly generated number?

Reply:
Yes, CertificateSigningRequests cannot be "mutated" after they are created. Therefore, to prevent previous failures from blocking the refresh-certs process, we use this seed to differentiate the names across attempts.

This can be just a random number. The only reason why we do this, is to know the names of the created CSRs that will need to be signed (and print the relevant message)